### PR TITLE
Replaces Traitor Greentext Objectives with Assassination/Escape Objectives

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
- Traitors now have the old two kill + escape objectives from before progressive traitors. These objectives are for greentext/victory conditions only, and are known as "Primary Objectives".
- The Progressive Traitor system remains in place for unlocking dangerous gear.
- Completing the Final Objective from the progressive list is an alternate route to victory.
- Traitor Uplinks now have a new tab that lists Primary Objectives, which will also indicate when the Traitor has a secured victory by Final Objective.
- Also cleaned up the round end report, see below.
## Why It's Good For The Game

See https://hackmd.io/WVjTLp6xTJiTFEkhSqqWIQ?view for my complete thoughts.

TL;DR:
- Getting Rep is too much of a time commitment for Greentext. 
- Most of ProgTraitor is quiet sabotage that no one cares about, leading to less conflict with the crew.
- Losing the Uplink creates an unwinnable state.

This PR fixes the above, by bringing back the 2 kill + escape system. Traitors are now required to actually negatively affect at least some crewmembers, and actually survive the round, to gain a victory state. Alternatively, if they reach the Final Objective, they are given a free pass, as an alternate route if they cannot find their Primary Objective targets. Final Objectives will count as a victory even if the traitor dies.

The Progressive Traitor system is still used for unlocking gear, and no values for items, or restrictions based on role, have been changed.
## Changelog
:cl:
balance: Traitors now have two kill and an escape objective for Greentext purposes, though completing a Final Objective will Greentext too.
/:cl:
New Primary Objective Tab:
tot_objs
If the Final Objective is done, the traitor's victory condition is secured, and the Primary Objectives no longer matter.
tot_objs_complete
Round-end report, showing one traitor that greentexts through traditional means, and one that doesn't get their kill target or escape, but greentexts via Final Objective anyway.
round_end

### CODE REVIEW

**Comment by Ryusenshu** [comment]
This might actually get me to enjoy traitor again
Thank you

**Comment by Serxule** [comment]
orange text doesnt really fit imo

**Comment by rasepretep** [comment]
I love this, but I do have a couple of questions:
 - does the assassinate objective work exactly like before, in that it unchecks itself if the target is revived?
 - is it only kill/escape objectives, no chance of rolling other ones, like stealing a sliver or the SD core?
 - not a question, but interested to see how this will impact lowpop rounds where dynamic tends to make 3/4 of the station traitors

Either way, amazing.

**Comment by carshalash** [comment]
I liked marooning objectives...

**Comment by zxaber** [comment]
- The assassinate objectives requires the target be dead at round-end. They won't mark themselves as complete mid-round (outside the Final Objective text).
- Currently it's only kill and escape. Well, Destroy AI has a roughly equal change of (Active AIs) / (Number of Living). I might add in maroon at a later date, but I don't plan on adding theft or glorious death objectives.

**Comment by rasepretep** [comment]
> * The assassinate objectives requires the target be dead at round-end. They won't mark themselves as complete mid-round (outside the Final Objective text).
> 
>     * Currently it's only kill and escape. Well, Destroy AI has a roughly equal change of (Active AIs) / (Number of Living). I might add in maroon at a later date, but I don't plan on adding theft or glorious death objectives.

I respect this. On the subject of maroon, it's a question of whether that even needs to be added, since most "prevent escape" objectives are simply done by killing the target. But maybe I'm entirely wrong and they require the target to actually be alive round-end?

Either way, I like this PR. And I'm saying this as someone who plays antag extremely rarely, hates being inconvenienced and despises being round removed. We do need this PR to break the monotony of nonstop boring kill pet/break lathe/bug office meta that made playing security a nightmare.

My only half-concern is how it'll play out on lowpop where dynamic tends to make 3/4 of the station traitors because it can't roll anything else. But I suppose it'll boil down to the playerbase itself and lowpop isn't really a valid way to measure it.

**Comment by Jacquerel** [comment]
I liked this better as originally proposed where it used the existing assassination objectives (behead/calling card) and also rewarded you for finishing them.  
If you’re just going back to old fashioned objectives can you at least make the escape one variable to escape/survive/hjack/martyr? That should be trivial.

**Comment by Ryusenshu** [comment]
> 
> I respect this. On the subject of maroon, it's a question of whether that even needs to be added, since most "prevent escape" objectives are simply done by killing the target. But maybe I'm entirely wrong and they require the target to actually be alive round-end?
> 


Maroon Objective was completed by either them staying on station or dead
Trap them somehow and get your greentext that way (emag shuttle before leaving / bolt it down / etc.)

It would be a nice addition just for the fact that not every evil plan has to involve killing with a desword

**Comment by zxaber** [comment]
>On the subject of maroon, it's a question of whether that even needs to be added, since most "prevent escape" objectives are simply done by killing the target. But maybe I'm entirely wrong and they require the target to actually be alive round-end?

Usually maroon was accomplished with a regular murder. But there were alternatives, such as tricking your target to doing something like saving the AI and then emagging the shuttle after they left. Generally, though, it came down to an easier version of assassinate, since rather than tracking down the target you'd just sweep the shuttle and hope they didn't catch a pod.

--

>I liked this better as originally proposed where it used the existing assassination objectives (behead/calling card) and also rewarded you for finishing them.

The main reason the two assassinate primary objectives don't give TC/rep is because it's hard to make a method of verifying them. This would be trivial on the PDA version of the uplink; just scan a dead body/head/brain with your PDA and it'll reward you. But the embedded uplink doesn't have that ability. 

>If you’re just going back to old fashioned objectives can you at least make the escape one variable to escape/survive/hjack/martyr? That should be trivial.

During one of my original conversations with Watermelon about this change, I believe I was told that we should not bring back the glorious death objective. But it's been a while, and I very well might be misremembering. Has happened before. Aside from that, I'd like to stick with Escape over Survive (as you have to at least hop onto the shuttle/a pod) but I don't have anything against Hijack. But that can be added in with a followup PR.

**Comment by Ryusenshu** [comment]
One question i have though is about the "Prog-kills"
Will those still be a thing?

Might be a bit too much if you still get kill objectives randomly in addition to your main one

**Comment by zxaber** [comment]
> One question i have though is about the "Prog-kills" Will those still be a thing?
> 
> Might be a bit too much if you still get kill objectives randomly in addition to your main one

Kill objectives do still show up as available for secondary/prog objectives. They specifically check to ensure you can't get a primary and secondary objective for the same person.

If you're looking to cut down on the murder, though, you can simply avoid the secondary ones.

**Comment by Bawhoppen** [comment]
This is good balance between the two systems

**Comment by KittyNoodle** [comment]
would it be possible for someone to get hijack/martyr objectives with this, because i would love to see those again

File: tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+          </Flex.Item>
+          <Flex.Item>
+            {primary_objectives.map((objective, index) => (
+              <Section
+                title={"Objective " + (index + 1)}
+                color={final_objective?"grey":"white"}>
+              {objective}
+              </Section>
+            ))}
+          </Flex.Item>
```
Comment (Watermelon914) [inline tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx:55]: Export and use the UI code from the traitor objective UI. Primary objectives should look like the secondary objective

File: tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+              position="absolute"
+              className="UplinkObjective__EmptyObjective" />
+            <NoticeBox
+              position="absolute"
+              width={"100%"}
+              fontSize="30px"
+              textAlign="center">
+              ALL OBJECTIVES MARKED COMPLETE 
+            </NoticeBox>
+          </Fragment>
```
Comment (Watermelon914) [inline tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx:33]: this notice box is really ugly, also I don't think "ALL OBJECTIVES MARKED COMPLETE" is a good phrase to use.
It may just be preferable to have nothing here, if not just displaying the final objective instead.

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Review by Watermelon914 — CHANGES_REQUESTED** [review]
Primary objectives panel has too much empty space, either integrate it into the 'Secondary Objective' tab or look into a way to fill out all of the empty space.

File: tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        )}
+        <Flex direction={'column'}>
+          <Flex.Item>
+            <Stack>
+              <Stack.Item
+                grow={1}
+                fill>
+                <Box mt={3} mb={3} bold fontSize={1.2} align="center" color="orange">
+                {'Agent, your Primary Objectives are as follows. Complete these at all costs.'}
```
Comment (Watermelon914) [inline tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx:42]: I don't really like how this looks at all, maybe just remove it completely?

File: code/modules/antagonists/traitor/uplink_handler.dm [hunk 4]
```diff
@@ -6,6 +6,8 @@
 /datum/uplink_handler
 	/// The owner of this uplink handler.
 	var/datum/mind/owner
+	/// The antag datum that we are linked to.
+	var/datum/antagonist/traitor/traitor_datum
```
[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/modules/antagonists/traitor/uplink_handler.dm:10]: sure, it's fine for easier access

[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/modules/antagonists/traitor/uplink_handler.dm:10]: It's just important to remember that they aren't necessarily a traitor.

**Comment by Troxonater** [comment]
I support this idea, Traitor used to be so much more fun when you HAD to do some objectives for the greentext, even if they were inconvenient. It forces players to get creative with how they do things if they're chasing the greentext. Also from a youtubers perspective I enjoyed structuring a video around work towards a goal the entire shift. This is the best of both worlds IMO with barely any compromise in design philosophy from old traitor or prog traitor.

**Comment by zxaber** [comment]
Implemented most of MrMelbert's suggestions.

Primary Objective screen now looks like the following;
image
-
image
The primary objectives now use the same Objective_Element as the secondary ones, abet stretched out because 3 doesn't divide into rows of 2 very well. The colors are done by setting all primary objectives to one color, except the last. This is somewhat hacky, and the colors will be incorrect if an admin adds an objective or does anything to make Escape not be the final one, but otherwise it should be fine.

I also added a check to the Objective_Element code so that the 0TC, 0 Reputation line is missing from final objectives (and thus primary ones too). To keep the proper height of the bottom of the element, a whitespace character is used.

File: code/modules/antagonists/traitor/objectives/assassination.dm [hunk 5]
```diff
@@ -171,6 +171,14 @@
 
 /datum/traitor_objective/assassinate/generate_objective(datum/mind/generating_for, list/possible_duplicates)
 
+	var/list/already_targeting = list() //List of minds we're already targeting. The possible_duplicates is a list of objectives, so let's not mix things
+	for(var/datum/objective/task as anything in handler.traitor_datum?.objectives)
+		if(!task.target)
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/assassination.dm:176]: This is redundant because of your istype check

File: code/modules/antagonists/traitor/objectives/eyesnatching.dm [hunk 6]
```diff
@@ -50,6 +50,15 @@
 	heads_of_staff = TRUE
 
 /datum/traitor_objective/eyesnatching/generate_objective(datum/mind/generating_for, list/possible_duplicates)
+
+	var/list/already_targeting = list() //List of minds we're already targeting. The possible_duplicates is a list of objectives, so let's not mix things
+	for(var/datum/objective/task as anything in handler.traitor_datum?.objectives)
+		if(!task.target)
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/eyesnatching.dm:56]: Same here

File: code/modules/antagonists/traitor/datum_traitor.dm [hunk 7]
```diff
@@ -47,6 +50,7 @@
 			uplink.uplink_handler = uplink_handler
 		else
 			uplink_handler = uplink.uplink_handler
+		uplink_handler.traitor_datum = src
```
Comment (Watermelon914) [inline code/modules/antagonists/traitor/datum_traitor.dm:53]: I think I would prefer you to not do this and to instead hold a list of permanent objectives on the uplink handler instead, which is then linked to the objectives list.

It would be better to keep uplink handlers completely independent from the antag datum, and to let the antag datum handle the uplink handler instead.

File: code/modules/antagonists/traitor/uplink_handler.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/modules/antagonists/traitor/uplink_handler.dm:10]: continuing on from this point, I would prefer for this to be a list of permanent objectives set by the traitor datum rather than a link to the traitor datum, for modularity purposes.

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            telecrystalPenalty={0}
+            progressionReward={0}
+            originalProgression={0}
+            HideTcRep={1}
+            canAbort={false}
+            grow={false}
+            finalObjective={false}
+          />
+        ))}
+      </Section>
```
Comment (Watermelon914) [inline tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx:75]: I'd like it if these were in a Stack element set to be vertical, so that there is a decent amount of spacing between each.

File: tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            <br />
+            CONNECTION CLOSED_
+          </Box>
+        </Dimmer>
+      )}
+      <Section>
+        {primary_objectives.map((prim_obj, index) => (
+          <ObjectiveElement
+            key={prim_obj.id}
+            name={'Objective ' + (index + 1)}
```
Comment (Watermelon914) [inline tgui/packages/tgui/interfaces/Uplink/PrimaryObjectiveMenu.tsx:55]: Calling them Objective 1/2/3 is a bit bland. Could you go for some different naming theme, e.g. 
"First Directive", "Second Directive" or something along those lines.

**Comment by zxaber** [comment]
UI now looks as such;

image

Using Phonetic Alphabet because the number of objectives is a config item, so I can't hardcode it. But if someone makes a Traitor have 27 Primary Objectives, it'll start spitting out numbers again.

It also uses `<Stack>` now.

File: code/datums/components/uplink.dm [hunk 10]
```diff
@@ -168,19 +168,28 @@
 	if(uplink_handler.has_objectives)
 		var/list/primary_objectives = list()
 		for(var/datum/objective/task as anything in uplink_handler.traitor_datum?.objectives)
-			primary_objectives += task.explanation_text
+			var/list/task_data = list()
+			if(length(primary_objectives) > length(GLOB.phonetic_alphabet))
+				task_data["task_name"] = "DIRECTIVE [primary_objectives.len]" //The english alphabet is WEAK
```
Comment (Fikou) [inline code/datums/components/uplink.dm:173]: why not use len + 1 if you were using it in the other one. also we prefer length(list) to list.len

**Comment by zxaber** [comment]
Primary Objective tab currently looks like this;
image

File: code/datums/components/uplink.dm (hunk 10 above)
Comment (zxaber) [reply code/datums/components/uplink.dm:173]: Oversight, thanks. Has been fixed.

File: code/modules/antagonists/traitor/uplink_handler.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/modules/antagonists/traitor/uplink_handler.dm:10]: Changed to using a list on the uplink handler that points to the same objective list the traitor datum uses.
